Fix PaymentGateway/Payment field handling

petebarnett opened the pull request:
Brings PaymentGateway bundle functionality into its own class.
Invokes `onBundleDelete` before the config entity is deleted, so that the plugin can be instantiated.
Fixes bugs with `hook_entity_bundle_delete` implementation and field api.

bojanz committed:
Issue #2711013: Implement the initial payment functionality

Commit:
Remove dependency on ConfigEntityBundleBase

Move payment gateway bundle functionality into its own class

Commit:
Fix bundle delete hook implementation

Check if field is still in use before deleting storage definition.
Fix bug where storage was deleted before bundle field itself

joshmiller83 commented on modules/payment/commerce_payment.module:
```diff
@@ -125,12 +139,16 @@ function commerce_payment_create_field_definitions(array $definitions) {
  * @param array $definitions
  *   The field definitions.
+ * @param bool $delete_storage
+ *   If set TRUE, field storage definition will be deleted also
```
But does it delete the payment entities? What are the ramifications for viewing past orders and payment entities?

petebarnett replied:
I guess it would reference a non-existent payment gateway
The scope of this PR was just to fix the field functionality associated with payments, but I agree that the above needs to be addressed at some point

joshmiller83 commented on modules/payment/src/Entity/PaymentGateway.php:
```diff
@@ -159,4 +161,67 @@ protected function getPluginCollection() {
+  /**
+   * {@inheritdoc}
+   */
+  public function postSave(EntityStorageInterface $storage, $update = TRUE) {
```
I feel like the order in the code should be: 
preSave
postSave
preDelete
postDelete

No?

joshmiller83 commented on modules/payment/src/Entity/PaymentGateway.php:
```diff
@@ -159,4 +161,67 @@ protected function getPluginCollection() {
+      \Drupal::getContainer()->get('entity_bundle.listener')->onBundleCreate($this->id(), $bundle_of);
+    }
+    else {
+      $entity_field_manager = \Drupal::getContainer()->get('entity_field.manager');
```
Can't we do this static and coupled ::getContainer() using a decoupled dependency injection? Or am I crazy?

bojanz replied:
We should be using \Drupal::service() here. Nobody uses \Drupal::getContainer() :)

Commit:
Update method order

Commit:
Use Drupal::service instead of getContainer

Commit:
use Drupal::service for entity_field.manager

bojanz commented on modules/payment/src/Entity/PaymentGateway.php:
```diff
@@ -159,4 +161,67 @@ protected function getPluginCollection() {
+  /**
+   * {@inheritdoc}
+   */
+  public static function postDelete(EntityStorageInterface $storage, array $entities) {
```
Why do we have an empty postDelete method?

petebarnett replied:
It's not empty, just calls parent... The only thing the original parent did was invoke the hook, which we discussed is no longer required.

bojanz commented on modules/payment/commerce_payment.module:
```diff
@@ -125,12 +139,16 @@ function commerce_payment_create_field_definitions(array $definitions) {
+ *   If set TRUE, field storage definition will be deleted also
  */
-function commerce_payment_delete_field_definitions(array $definitions) {
+function commerce_payment_delete_field_definitions(array $definitions, $delete_storage=FALSE) {
```
Missing space around '=' for $delete_storage

Commit:
code style fix

petebarnett commented on modules/payment/commerce_payment.module:
```diff
@@ -125,12 +139,16 @@ function commerce_payment_create_field_definitions(array $definitions) {
   $field_storage_definition_listener = \Drupal::service('field_storage_definition.listener');
   $field_definition_listener = \Drupal::service('field_definition.listener');
   foreach ($definitions as $definition) {
-    $field_storage_definition_listener->onFieldStorageDefinitionDelete($definition);
```
Ignoring the big change, there is a subtle bug-fix in here as calling `onFieldStorageDefinitionDelete` before `onFieldDefinitionDelete` would result in database errors as Drupal tries to update tables that no longer exist.

Commit:
Fix bug in converting from EntityManager

clearCachedBundles method exists on EntityTypeBundleInfo, not EntityFieldManager

bojanz commented:
I've abstracted out the field handling into a generic API: https://github.com/drupalcommerce/commerce/commit/e09ca1770018dfea1593340e489a36878e3cd7d0
Will be used by Promotions as well, and later moved to Entity API.

After doing that, and looking into this PR, I realized that the gateway approach for bundles that we have here is too magical, so I introduced PaymentType plugins, linked from the gateway instead. 

Thus, I am closing this PR, and will credit you on one of the followup commits, to make up for it.
